Blk

jankara opened the pull request:
Hi Jeff!

These changes implement 'blk -d' command.

jankara committed:
crash.subsystem.storage: Cleanup singlequeue definitions

Rename function definitions in blocksq.py to start with sq prefix so
that we can have also corresponding mq alternatives without clashing
names. Also move request_age_ms to common part since it is not sq
specific.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara committed:
crash.subsystem.storage: Add function to test for blk-mq queues

Add function testing whether a given queue is a multiqueue queue or not.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara committed:
crash.subsystem.storage: Add checks for queue type in sq functions

Add checks to functions for single-queue queues that they are indeed
operating on such queues.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara committed:
crash.subsystem.storage: Add blk-mq support

Add support for iterating requests in blk-mq queue and reporting number
of requests in flight. Also provide generic wrappers detecting queue
type and using appropriate single / multi queue handlers.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara committed:
crash.subsystem.storage: Add function to query number of queued request

Add function to query number of queued requests for single queue request
queues.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara committed:
crash.subsystem.storage: Add function to query all queue request stats

Since gathering of blk-mq queue statistics about requests is relatively
expensive (we have to iterate all tag bitmaps and inspect all requests),
provide a function to gather all information at once so that we don't
have to iterate everything multiple times.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara committed:
crash.commands: Add dev command

Add 'dev -d' command.

Signed-off-by: Jan Kara <jack@suse.cz>

jeffmahoney commented on crash/subsystem/storage/blocksq.py:
```diff
@@ -14,7 +13,7 @@ class NoQueueError(RuntimeError):
 types = Types(['struct request'])
 
-def for_each_request_in_queue(queue: gdb.Value) -> Iterable[gdb.Value]:
+def sq_for_each_request_in_queue(queue: gdb.Value) -> Iterable[gdb.Value]:
```
I don't have a problem with renaming it, but we should retain the alias. Code that wants to use both interfaces can be changed to use the prefix but this breaks existing scripts.  It's possible to use both, even with the old name, in the same script by doing e.g. import crash.subsystem.storage.blocksq as blocksq and then using blocksq.for_each_request_in_queue.

jankara replied:
I agree this commit breaks it but following commits do add for_each_request_in_queue() again which is then properly distinguishing sq and mq queues and calling appropriate functions. That being said I've now realized the function will come from a different module (crash/subsystem/storage/block.py not blocksq.py) so scripts using this will indeed be broken. But won't it be better to fix scripts - IHMO we should have the external block layer API in block.py and leave blocksq and blockmq for low level stuff of sq / mq queues. What do you think?

jeffmahoney commented on crash/subsystem/storage/blocksq.py:
```diff
@@ -34,24 +33,7 @@ def for_each_request_in_queue(queue: gdb.Value) -> Iterable[gdb.Value]:
-    return jiffies_to_msec(kernel.jiffies - request['start_time'])
-
-def requests_in_flight(queue: gdb.Value) -> Tuple[int, int]:
+def sq_requests_in_flight(queue: gdb.Value) -> Tuple[int, int]:
```
Same here.

jeffmahoney commented on crash/subsystem/storage/__init__.py:
```diff
@@ -263,6 +275,49 @@ def request_age_ms(request: gdb.Value) -> int:
+        :obj:`bool`: True for synchronous requests, False otherwise.
+    """
+    return request['cmd_flags'] & 1 == 0 or \
+           request['cmd_flags'] & (REQ_SYNC | REQ_FUA | REQ_PREFLUSH) != 0
```
Use parens instead of \ here and in the next few stanzas.

jankara replied:
Aha, that's the python trick ;) Thanks.

jeffmahoney reviewed (changes requested):
There are a few minor nits to clean up and two bigger things: removing the old name for for_each_request_in_queue and handling the enums.  I could be convinced to accept the enums as-is if they pass tests (incl. static-check and lint).

Thanks,

-Jeff

Update: The enum loading doesn't past testing:

crash.exceptions.DelayedAttributeError: Delayed attribute enum_req_flag_bits_type has not been completed.

jankara committed:
contrib: Fixup stuck-requests.py

Fixup stuck-requests.py to import for_each_request_in_queue from block
and not blocksq where it used to be.

Signed-off-by: Jan Kara <jack@suse.cz>

jankara commented:
I went trough results of 'make test' and fixed up some problems. What is still remaining is:
crash/subsystem/storage/__init__.py:298: error: Name 'REQ_SYNC' is not defined
crash/subsystem/storage/__init__.py:298: error: Name 'REQ_FUA' is not defined
crash/subsystem/storage/__init__.py:298: error: Name 'REQ_PREFLUSH' is not defined
crash/subsystem/storage/__init__.py:368: error: Name 'REQ_STARTED' is not defined

These are autodefined flags static checker has no idea about. Do you know how to silence it?

crash/subsystem/storage/block.py:84: error: Incompatible return value type (got "Tuple[int, ...]", expected "Tuple[int, int, int, int]")

This is because type checker is not able to infer that [int, int] + [int, int] == [int, int, int, int]. I can work this around by constructing the tuple like (a[0], a[1], b[0], b[1]) but is it worth it?

jankara commented:
OK, resolved the remaining issues by silencing the checker. I'll close this request and create a new one.
